### PR TITLE
Latest_stable_version

### PR DESCRIPTION
**Pull request by ConstVol**
- sorting csv IMDB data by movie genre
- inline keyboard to choose movie genre
- rutracker API integration
  gem 'rutracker_api' out of date - need to change login link.
- cool oneliner functions

Probably oneliners are difficult to read but i wanted to try it.

### CODE REVIEW

File: bot.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+require 'telegram/bot'
+require 'csv'
+require 'rutracker_api'
+
+class IMDB_Helper
+
+  size_to_kb = {'KB' => 1, 'MB' => 1024, 'GB' => 1024*1024} # convert everything to KB in order compare sizes
+  client = RutrackerApi.new('username', 'password')
```
Comment (anatoliliotych) [inline bot.rb]: where do you store username and password?

File: bot.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  csv = CSV.new(File.read('./data/movie_metadata.csv'), :headers => true)
+
+  # sorting IMDB data in {'genre' => [{'title' => 'imbd_link'}, ...], ...} form. In one line just for fun;)
+  csv.each {|row| row.to_hash['genres'].split('|').each {|genre| genres[genre].push({row.to_hash['movie_title'] => row.to_hash['movie_imdb_link']})} if row.to_hash['imdb_score'].to_f > 7}
+
+  # create array of telegram buttons, one for each genre
+  buttons = []
+  genres.keys.each { |genre| buttons << Telegram::Bot::Types::InlineKeyboardButton.new(text: genre, callback_data: genre)}
+
+  Telegram::Bot::Client.run('telegram_bot_token') do |bot|
```
Comment (anatoliliotych) [inline bot.rb]: what is the way to set token?

File: bot.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+require 'telegram/bot'
+require 'csv'
+require 'rutracker_api'
+
+class IMDB_Helper
```
Comment (anatoliliotych) [inline bot.rb]: how do you use this class? how to launch bot?

File: bot.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  size_to_kb = {'KB' => 1, 'MB' => 1024, 'GB' => 1024*1024} # convert everything to KB in order compare sizes
+  client = RutrackerApi.new('username', 'password')
+
+  # sort imdb data by movie genre
+  genres = Hash.new { |hash, key| hash[key] = [] } # needed to avoid .has_key? check
+  csv = CSV.new(File.read('./data/movie_metadata.csv'), :headers => true)
+
+  # sorting IMDB data in {'genre' => [{'title' => 'imbd_link'}, ...], ...} form. In one line just for fun;)
+  csv.each {|row| row.to_hash['genres'].split('|').each {|genre| genres[genre].push({row.to_hash['movie_title'] => row.to_hash['movie_imdb_link']})} if row.to_hash['imdb_score'].to_f > 7}
```
Comment (anatoliliotych) [inline bot.rb]: too long string! try to make code more readable

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

File: bot.rb (hunk 3 above)
Comment (ConstVol) [reply bot.rb]: I don't use it. I'll get rid of it.

File: bot.rb (hunk 1 above)
Comment (ConstVol) [reply bot.rb]: I'll create constants on top

File: bot.rb (hunk 4 above)
Comment (ConstVol) [reply bot.rb]: will revert to a separete method

File: bot.rb (hunk 2 above)
Comment (ConstVol) [reply bot.rb]: I'll create constant on top

Comment (anatoliliotych) [reply bot.rb]: use .env file for this with gem dotenv

File: bot.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
+require 'telegram/bot'
+require 'csv'
+require 'rutracker_api'
+
+USERNAME = ''
```
Comment (anatoliliotych) [inline bot.rb]: will this code work? or you will need to put creds here before launch? :)

use .env file with dotenv gem

File: bot.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+def sort_by_genres
+
+  genres = Hash.new { |hash, key| hash[key] = [] } # needed to avoid .has_key? check
+  csv = CSV.new(File.read('./data/movie_metadata.csv'), :headers => true)
+
+# sorting IMDB data in {'genre' => [{'title' => 'imbd_link'}, ...], ...} form
+  csv.each do |row|
+    movie = row.to_hash
+    movie['genres'].split('|').each do|genre|
```
Comment (anatoliliotych) [inline bot.rb]: style guide!

File: bot.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+      if movie['imdb_score'].to_f > 7
```
Comment (anatoliliotych) [inline bot.rb]: to float and comparing with integer

File: bot.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  bot.listen do |message|
+
+    case
+      when message.is_a?(Telegram::Bot::Types::CallbackQuery)
+
+        movie = genres[message.data].sample # get random movie by genre in ['title' => 'imdb_link'] form
+        search_result = client.search(term:movie.keys[0])
+
+        #searching on rutracker by movie title(checking if response isn't {:error => 'Not found'}) -> sorting response by size -> getting the last one(biggest one)
+        torrent_id = client.search(term:movie.keys[0]).sort_by{|torrent| torrent[:size].split(" ")[0].to_f * size_to_kb[torrent[:size].split(" ")[1]]}.last[:torrent_id] unless search_result.is_a? Hash
```
Comment (anatoliliotych) [inline bot.rb]: can you read this? try to make code readable

File: bot.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+size_to_kb = {'KB' => 1, 'MB' => 1024, 'GB' => 1024*1024} # convert everything to KB in order compare sizes
+client = RutrackerApi.new(USERNAME, PASSWORD)
+
+genres = sort_by_genres # sort imdb data by movie genre
+# create array of telegram buttons, one for each genre
+buttons = []
+genres.keys.each { |genre| buttons << Telegram::Bot::Types::InlineKeyboardButton.new(text: genre, callback_data: genre)}
+
+Telegram::Bot::Client.run(TOKEN) do |bot|
+
```
Comment (anatoliliotych) [inline bot.rb]: remove empty lines 
read style guide about this

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

File: bot.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+sorter = Sorter.new
+rutracker_handler = RutrackerHandler.new
+genres = sorter.sort_by_genres # sort imdb data by movie genre
+buttons = [] # create array of telegram buttons, one for each genre
+genres.keys.each { |genre| buttons << Telegram::Bot::Types::InlineKeyboardButton.new(text: genre, callback_data: genre) }
+Telegram::Bot::Client.run(ENV['TOKEN']) do |bot|
+  bot.listen do |message|
+    case
+      when message.is_a?(Telegram::Bot::Types::CallbackQuery)
+        movie = genres[message.data].sample # get random movie by genre in ['title' => 'imdb_link'] form
```
Comment (anatoliliotych) [inline bot.rb:15]: If comments are obvious, I'd skip them.

File: rutracker_api.rb [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  # @return [Hash] the options keys with search result
+  def search(options = {})
+    query = prepare_query_string options
+    @agent.get query
+    parse_search
+  end
+
+  def find_user(nick)
+    # TODO: add parser for user pages
+    @agent.get 'http://rutracker.org/forum/profile.php?mode=viewprofile&u=' + nick
```
Comment (anatoliliotych) [inline rutracker_api.rb]: URL to constant! you have lots of common in urls, try to prepare method which removes duplication

File: rutracker_api.rb [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+  end
+
+  private
+  def login(username, pass)
+    @agent.post(LOGIN_PAGE, login_username: username,
+                login_password: pass, login: 'Вход')
+
```
Comment (anatoliliotych) [inline rutracker_api.rb]: remove empty line

File: rutracker_api.rb [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    end
+    query << "&f=#{prepared[:f]}" if prepared[:f]
+    query << "&f=#{prepared[:o]}" if prepared[:order_by]
+    query << "&f=#{prepared[:s]}" if prepared[:sort]
+    query
+  end
+
+  def parse_search
+    @agent.page.search("//table[@class='forumline tablesorter']/tbody/tr").map do |row|
+      if row.css(".row1").text =~ /Не найдено/
```
Comment (anatoliliotych) [inline rutracker_api.rb]: make oneliner

File: rutracker_handler.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    @client = RutrackerApi.new(ENV['USERNAME'], ENV['PASSWORD'])
+  end
+
+  def get_torrend_url(movie)
+    sorter = Sorter.new
+    search_result = @client.search(term:movie.keys[0])
+    unless search_result.is_a? Hash
+      sorted_by_size = sorter.sort_by_torrent_size(search_result)
+      torrent_id = sorted_by_size.last[:torrent_id]
+      "https://rutracker.org/forum/viewtopic.php?t=#{torrent_id}"
```
Comment (anatoliliotych) [inline rutracker_handler.rb]: url to constant

File: sorter.rb [hunk 15]
```diff
@@ -0,0 +1,33 @@
+require 'csv'
+
+class Sorter
+
+  def initialize
+    @file_name = './data/movie_metadata.csv'
```
Comment (anatoliliotych) [inline sorter.rb]: filename should be in constant

File: sorter.rb [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    @size_to_kb = {'KB' => 1, 'MB' => 1024, 'GB' => 1024*1024} # convert everything to KB in order compare sizes
```
Comment (anatoliliotych) [inline sorter.rb]: use constant

Comment (anatoliliotych) [inline sorter.rb]: check style

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

File: rutracker_api.rb [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+class RutrackerApi
+  attr_accessor :agent
+
+  LOGIN_PAGE = 'http://rutracker.org/forum/login.php'
+  SEARCH_PAGE = 'http://rutracker.org/forum/tracker.php'
+
+  ORDER_OPTIONS = { date: 1,  name: 2, downloads: 4, shows: 6, seeders: 10,
+                    leechers: 11, size: 7, last_post: 8, speed_up: 12,
+                    speed_down: 13, message_count: 5, last_seed: 9 }
+  SORT_OPTIONS = {asc: 1, desc: 2}
```
Comment (anatoliliotych) [inline rutracker_api.rb]: style guide!

File: rutracker_api.rb [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    query << "&f=#{prepared[:f]}" if prepared[:f]
+    query << "&f=#{prepared[:o]}" if prepared[:order_by]
+    query << "&f=#{prepared[:s]}" if prepared[:sort]
+    query
+  end
+
+  def parse_search
+    @agent.page.search("//table[@class='forumline tablesorter']/tbody/tr").map do |row|
+      if row.css(".row1").text =~ /Не найдено/
+        return {error: 'Not found'}
```
Comment (anatoliliotych) [inline rutracker_api.rb]: style guide

**Review by anatoliliotych — COMMENTED** [review]
respond to code review!

File: bot.rb (hunk 10 above)
Comment (ConstVol) [reply bot.rb:15]: For some bodies it may not be so obvious

File: rutracker_api.rb (hunk 13 above)
Comment (ConstVol) [reply rutracker_api.rb]: It would be disrespectful to the creator of rutracker_api gem:)

File: bot.rb (hunk 10 above)
Comment (anatoliliotych) [reply bot.rb:15]: try to write obvious and readable code :)

File: rutracker_api.rb (hunk 13 above)
Comment (anatoliliotych) [reply rutracker_api.rb]: please make : 
```
return {} if condition
```

File: bot.rb [hunk 19]
```diff
@@ -0,0 +1,24 @@
+require 'telegram/bot'
+require 'dotenv/load'
```
Comment (anatoliliotych) [inline bot.rb:2]: add .env.example file

File: rutracker_api.rb [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+  def login(username, pass)
+    @agent.post(LOGIN_PAGE, login_username: username,
+                login_password: pass, login: 'Вход')
+  end
+
+  def prepare_query_string(options)
+    prepared = { f: options[:category], nm: options[:term],
+                 s: SORT_OPTIONS[options[:sort]], o: ORDER_OPTIONS[options[:order_by]] }
+
+    query = SEARCH_PAGE + '?'
```
Comment (anatoliliotych) [inline rutracker_api.rb]: is it worth to have this + ?

**Review by anatoliliotych — CHANGES_REQUESTED** [review]

**Review by anatoliliotych — CHANGES_REQUESTED** [review]
Fix commit messages

**Review by anatoliliotych — APPROVED** [review]
LGTM

**Pull request merged**